src/Transit/Internal/Pipeline.hs

```diff
@@ -0,0 +1,132 @@
+sendPipeline fp (TCPEndpoint s) key =
+  C.sourceFile fp .| sha256PassThroughC `C.fuseBoth` (encryptC key .| CN.sinkSocket s)
+
+-- | Receive the encrypted bytestream from a network socket, decrypt it and
+-- write it into a file, also calculating the sha256 sum of the decrypted
+-- output along the way.
+receivePipeline :: C.MonadResource m =>
+                   FilePath
+                -> Int
+                -> TCPEndpoint
+                -> SecretBox.Key
+                -> C.ConduitM a c m (Text, ())
+receivePipeline fp len (TCPEndpoint s) key =
+    CN.sourceSocket s
+    .| assembleRecordC
+    .| decryptC key
+    .| CB.isolate len
+    .| sha256PassThroughC `C.fuseBoth` C.sinkFileCautious ("./" </> fp)
+
+encryptC :: Monad m => SecretBox.Key -> C.ConduitT ByteString ByteString m ()
+encryptC key = go Saltine.zero
+  where
+    go nonce = do
+      b <- C.await
+      case b of
+        Nothing -> return ()
+        Just chunk -> do
+          let cipherText = encrypt key nonce chunk
+              cipherTextSize = toLazyByteString (word32BE (fromIntegral (BS.length cipherText)))
+          C.yield (toS cipherTextSize)
+          C.yield cipherText
+          go (Saltine.nudge nonce)
+
+decryptC :: MonadIO m => SecretBox.Key -> C.ConduitT ByteString ByteString m ()
+decryptC key = loop
+  where
+    loop = do
+      b <- C.await
+      case b of
+        Nothing -> return ()
+        Just bs -> do
+          let (nonceBytes, ciphertext) = BS.splitAt boxNonce bs
+              nonce = fromMaybe (panic "unable to decode nonce") $
+                Saltine.decode nonceBytes
+              maybePlainText = SecretBox.secretboxOpen key nonce ciphertext
+          case maybePlainText of
+            Just plaintext -> do
+              C.yield plaintext
+              loop
+            Nothing -> throwIO (CouldNotDecrypt "SecretBox failed to open")
+
+sha256PassThroughC :: (Monad m) => C.ConduitT ByteString ByteString m Text
+sha256PassThroughC = go $! Hash.hashInitWith SHA256
+  where
+    go :: (Monad m) => Hash.Context SHA256 -> C.ConduitT ByteString ByteString m Text
+    go ctx = do
+      b <- C.await
+      case b of
+        Nothing -> return $! show (Hash.hashFinalize ctx)
+        Just bs -> do
+          C.yield bs
+          go $! Hash.hashUpdate ctx bs
+
+-- | The decryption conduit computation would succeed only if a complete
+-- bytestream that represents an encrypted block of data is given to it.
+-- However, the upstream elements may chunk the data for which one may not
+-- have control of. The encrypted packet on the wire has a 4-byte length
+-- header, so we could first read it and assemble a complete encrypted
+-- block into downstream.
+assembleRecordC :: Monad m => C.ConduitT ByteString ByteString m ()
+assembleRecordC = do
+  b <- C.await
+  case b of
+    Nothing -> return ()
+    Just bs -> do
+      let (hdr, pkt) = BS.splitAt 4 bs
+      let len = runGet getWord32be (BL.fromStrict hdr)
```

Comment:
ooh, that's right. We can end up in a situation where the upstream yields less than 4 bytes and then `runGet` would blow up.

The problem is that `runGet` is designed to be used as a non-sink element in a pipeline, so it returns a unit, but passes the chunk to the next downstream block, which means, we probably means, we cannot do a `header <- getChunk 4 BB.empty`. May be I will try another approach where if the yielded bytestring has a length less than 4, it will request a chunk of size 4 using `getChunk` and store the vestiges in the bytestring builder for the next iteration to consume..